### DISCUSSION TITLE
dnsmasq not working/starting with newest build 7590 07.57

### DISCUSSION THREAD

**Discussion by frag1x** [Freetz-NG/freetz-ng#839] [Allgemein] [answered]
Hi,

I just updated my Freetz image to the newest build freetz-ng-22375MF-4542d10bb for the 7590 7.57 from my old 7.50 freetz build. I didn't change anything in the configuration but nevertheless rerun "make menuconfig".

Everything works fine and the same like in my previous build, except from dnsmasq. The dnsmasq configuration entry in the webinterface is missing and the dnsmasq service is not shown under services in the webif.

Under Mod-Info, dnsmasq is listed under Packages. 

I tried to start the service manually via "rc.dnsmasq start" but get the error:

root@fritzbox:/etc/init.d# rc.dnsmasq start
Starting Dnsmasq ...
dnsmasq: failed to bind DHCP server socket: Address already in use
failed.

So it seems like there is a problem with the multid DNS/DHCP interfering with dnsmasq again in the new version. I selected "Disabled multid's DNS/DHCP" in "make menuconfig".

**Comment by wcdgit** [comment]
I have the same problem with my 7490 Box. I already did a clean git clone and startet the built from scratch with a new menuconfig. The problem persists. Even with a minimal config with only standard settings and dnsmasq I had no luck.

![grafik](https://github.com/Freetz-NG/freetz-ng/assets/54795276/7500b3e6-545e-4790-a552-86056400b7f7)

**Comment by fda77** [comment]
The menu entry on webif is missing when "rc.dnsmasq load" was not executed. This is done by multid wrapper very early. Maybe you have some not included webifs / addons?
7490 works for me

**Reply by frag1x** [reply]
Thanks, can you tell me what the name of the multid wrapper file is and the location where I can find it? So I can check if "rc.dnsmasq load" is present there.

If I execute "rc.dnsmasq load" manually I don't get any output, I don't know if this is normal:

root@fritzbox:/etc/init.d# rc.dnsmasq load
root@fritzbox:/etc/init.d#

**Comment by wschleifer** [comment]
I have the same problem. What do I have to do?

BTW: I've used a config that is working with DnsMasq and 7.56. but for 7.57 not.

Thanks.

**Comment by fda77** [comment]
7.56 and 7.57 have very few differences, which should not be related to dns. So you erroro repors dont make sense

**Comment by schacka** [comment]
Same problem here with the exact same error (rc.dnsmasq start --> dnsmasq: failed to bind DHCP server socket: Address already in use).
I have a 6591 box with 7.56. Box and version are likely not the cause. As far as i can remember, probs started with "[mod: allow disabling avm services](https://github.com/Freetz-NG/freetz-ng/commit/7baa978629981e63121c31ceecea6ca53a2a0d18)"

**Comment by frag1x** [comment]
So somehow, "rc.dnsmasq load" is not or not correctly executed on our Boxes. Are there any logfiles where this execution can be tracked? In [/var/log/mod.log] or [/var/log/mod_load.log] I don't see anything related to dnsmasq.

Edit: I found a forum post with the exact same error description "dnsmasq not loading/preset; dnsmasq Webif entry missing" from the 25th of august.
So either there is a timezone mismatch between github commit time (26th of august) and this forum post or the problem was already present before the "[mod: allow disabling avm services]" commit.

Edit: the version from the forum post is "ng-22301-93bac00d5"

**Comment by frag1x** [comment]
Update: I did some testing with old commits. The problem first occured after the commits from august 21th.

If you want to test it yourself: 

"git checkout 5874b508b4de94432c5bc44ca83a3e29e1b0a34c" dnsmasq is working (august 20th))

"git checkout f30b7265d1b25c1aff5c85c57e0c941634b3c3fd" dnsmasq is not working (august 21th)

There were changes on the multid wrapper at this time.

**Comment by frag1x** [comment]
I think I found the problem.

In the patchfile "106-wrapper_avm.sh" there is a line
```
for daemon in dsld multid rextd; do
[ -e "${FILESYSTEM_MOD_DIR}$wrapath/$daemon" ] || continue
```
which checks for the existence of a file (dsld, multid, rextd) under "build/modified/filesystem/usr/bin/wrapper/". But none of the files seem to be there at that time and the loop never continues for any of these. So the following steps won't execute and the "${FILESYSTEM_MOD_DIR}/lib/systemd/system/multid.service" file won't get patched.

You can check it on your freetz box: "less /lib/systemd/system/multid.service"
If the patch didn't run successfully, there is the line
```
[Service]
ExecStart=/sbin/multid
```
On a working system the line is:
```
ExecStart=/usr/bin/wrapper/multid
```
So if you remove the line
```
[ -e "${FILESYSTEM_MOD_DIR}$wrapath/$daemon" ] || continue
```
in the "/patches/scripts/106-wrapper_avm.sh" file, everything should work again. I just flashed the newst build with the edited file and dnsmasq is working for me again.

**Comment by fda77** [comment]
Thanks for your investigation, I think you write about 106aa7d4b9524b3b99d6ca2f47d9e5076dddd016
"dnsmasq does not go/run/work" "reports" are very old and maybe there are different problems as they are older than this commit.

**Reply by fda77** [reply, marked as answer]
* 1ffd6f30e dnsmasq: fix 106aa7d4b9524b3b99d6ca2f47d9e5076dddd016, by @frag1x

**Reply by schacka** [reply]
For me, the fix also solves the problem (6591 box). But now there is another one. DSLD can not assign ipv6 address --> reboot (watchdog)
Maybe there is also a dependency here that was overlooked. I'm not enough of an expert to solve it. With dnsmasq disabled (so no preload) everything works fine for now.

**Reply by fda77** [reply]
All the dnsmasq error reports dont make sense, example:
 - https://www.ip-phone-forum.de/threads/315112/post-2527565

this is the same as yours, but since 7.50! Now fixed by the commit here. But the fix here revers a commit ~2 weeks old.
So i cann not track all the "dont work" reports

As developer you could disable the watchdog in menuconfig to have more time

**Reply by schacka** [reply]
I went back to it. Everything works fine without problems now.

In my case it was an "onlinechanged" script, which creates a blocklist (curl), saves the list as a file (/tmp/<blocklist>) and then restarts dnsmasq (option addn-hosts=/tmp/<blocklist>). Disabling the script stopped the reboots.
As far as I can see, dsld establishes a connection to the cable network and dnsmasq is restarted afterwards. If the script gets in between something goes wrong. 

~~Using `sleep 10 && <script for list> && sleep 10 && /etc/init.d/rc.dnsmasq restart` as an onlinechanged entry works fine.~~

```
Update: My script fetching the blocklist was buggy :(  Now it's:
`curl $(cat >dnsmasq-adblock-list.txt> | grep -vE "^$|^#") 2>/dev/null | grep -vE "#|^$|http:|https:|0.0.0.0$|localhost$|localdomain$|local$|broadcasthost$|ip6-localhost$|ip6-loopback$|ip6-localnet$|ip6-mcastprefix$|ip6-allnodes$|ip6-allrouters$|ip6-allhosts$" | awk -F' ' '!a[$NF]++ {gsub(/^/,"0.0.0.0 ",$NF) ; print $NF ; gsub(/^0\.0\.0\.0/,"::1",$NF) ; print $NF}' > /tmp/dnsmasq.adblock.tmp`
`/etc/init.d/rc.dnsmasq stop`
`mv -f /tmp/dnsmasq.adblock.tmp /tmp/dnsmasq.adblock`
`/etc/init.d/rc.dnsmasq restart`

```

**Reply by fda77** [reply]
Lol 
is your connection up in 10 seconds? My (/a?) shitty telekom connections need about 2:50 minutes.
And Dnsmasq needs only a "reload" to apply new lists
Beside that, you should consider to use https://freetz-ng.github.io/freetz-ng/make/addhole.html

**Reply by schacka** [reply]
:) No, it's the onlinechanged - online script. So after 2:49 minutes i guess :P
Thanks for the reload info. Already tried addhole, but the above gives me a nice list for ipv4 and ipv6.
Sometimes reloading gives me this error:
`kern.info kernel: [45484.878778][0]dnsmasq[10427]: segfault at c ip 6e24f616 sp 7ffbca10 error 4 in libuClibc-1.0.44.so[6e21b000+58000]`

So i ended with `/etc/init.d/rc.dnsmasq reload || /etc/init.d/rc.dnsmasq restart`

Thank you for helping me out!

**Reply by fda77** [reply]
"normal" dsl training need here min 170 seconds, its crappier than average dsl
Why v4+v6 when you want block HOSTS ? It does not matter what's the target of redirected host, and with adhole you could change the target ip, v6 should also work but devices without v6 get nothing .
When dnsmasq crahes (?), attach strace (-p PID) and do again the reload action

**Reply by schacka** [reply]
```
`--- SIGHUP {si_signo=SIGHUP, si_code=SI_USER, si_pid=11363, si_uid=0} ---
getpid()                                = 6932
writev(12, [{iov_base="\1\0\0\0\0\0\0\0\0\0\0\0", iov_len=12}], 1) = 12
sigreturn({mask=[]})                    = -1 EINTR (Interrupted system call)
poll([{fd=4, events=POLLIN}, {fd=5, events=POLLIN}, {fd=6, events=POLLIN}, {fd=7, events=POLLIN}, {fd=8, events=POLLIN}, {fd=9, events=POLLIN}, {fd=10, events=POLLIN}, {fd=11, events=POLLIN}], 8, -1) = 1 ([{fd=11, revents=POLLIN}])
time(NULL)                              = 1694814166 (2023-09-15T23:42:46+0200)
read(11, "\1\0\0\0\0\0\0\0\0\0\0\0", 12) = 12
open("/etc/hosts", O_RDONLY|O_LARGEFILE) = 14
ioctl(14, TCGETS, 0x7f9edd78)           = -1 ENOTTY (Inappropriate ioctl for device)
brk(0x8add000)                          = 0x8add000
read(14, "127.0.0.1\tlocalhost\n192.168.20.1"..., 4096) = 60
read(14, "", 4096)                      = 0
close(14)                               = 0
getpid()                                = 6932
time([1694814166 /* 2023-09-15T23:42:46+0200 */]) = 1694814166 (2023-09-15T23:42:46+0200)
open("/etc/TZ", O_RDONLY)               = 14
read(14, "CET-1CEST-2,M3.5.0/02:00:00,M10."..., 68) = 45
close(14)                               = 0
getpid()                                = 6932
write(13, "<30>Sep 15 23:42:46 dnsmasq[6932"..., 60) = 60
munmap(0x42cdb896, 3391088841)          = -1 EINVAL (Invalid argument)
munmap(0x68064501, 4294967291)          = -1 EINVAL (Invalid argument)
stat64("/var/tmp/flash/dnsmasq/dnsmasq.dhcphosts", {st_mode=S_IFREG|0644, st_size=1291, ...}) = 0
open("/var/tmp/flash/dnsmasq/dnsmasq.dhcphosts", O_RDONLY|O_LARGEFILE) = 14
ioctl(14, TCGETS, 0x7f9eddc8)           = -1 ENOTTY (Inappropriate ioctl for device)
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0xc} ---
+++ killed by SIGSEGV +++`
```

**Comment by fda77** [comment]
> write(13, "<30>Sep 15 23:42:46 dnsmasq[6932"..., 60) = 60

Maybe you should read the logfile/syslog output

**Reply by schacka** [reply]
What do you mean by this? Which logfile? The only thing i found in /var/log/messages:
user.err multid[2222]: dhcp_reset_binary_options: iface lan not found
user.err cableinfo[2603]: di_avmipc_receive_handler: cannot map channel 0 to demuxer

**Reply by fda77** [reply]
> Maybe you should read the logfile/syslog output

**Reply by schacka** [reply]
I did. For example in strace:

write(13, "<30>Sep 17 18:15:11 dnsmasq-dhcp"..., 86) = 86

in syslog:
Sep 17 18:15:11 daemon.info dnsmasq[17858]: read /tmp/.dnsmasq.addhole - 200853 names
Sep 17 18:15:11 daemon.info dnsmasq-dhcp[17858]: read /var/tmp/flash/dnsmasq/dnsmasq.dhcphosts
Sep 17 18:15:11 kern.warn kernel: [150667.765156][1][1]system-load  42% loadavg 0.32 0.23 0.18 - task runtime:5% max:kworker/u4:2 0%  softirqs:3.50% (ce_tasklet [qca_ol] 72%) pgstat: sum=171106 free=45144 slab=12271 alloc=2708/s fault=3034/s (sleep 8)

What could be the problem?

**Reply by fda77** [reply]
The content of you file...
Try to add only some example lines and test if it then still crashes.
Btw, adhole removes invalid chars ;-)

**Reply by schacka** [reply]
It's crazy as hell. If i use adhole and clear the hostfile the content of the file looks like this:
`#<LF>`
```
Reloading dnsmasq gives me:

`getpid()                                = 11890
write(13, "<30>Sep 17 19:00:05 dnsmasq[1189"..., 61) = 61
stat64("/tmp/.dnsmasq.addhole", {st_mode=S_IFREG|0644, st_size=2, ...}) = 0
open("/tmp/.dnsmasq.addhole", O_RDONLY|O_LARGEFILE) = 14
ioctl(14, TCGETS, 0x7fc6af48)           = -1 ENOTTY (Inappropriate ioctl for device)
read(14, "#\n", 4096)                   = 2
read(14, "", 4096)                      = 0
close(14)                               = 0
getpid()                                = 11890
time([1694970005 /* 2023-09-17T19:00:05+0200 */]) = 1694970005 (2023-09-17T19:00:05+0200)
open("/etc/TZ", O_RDONLY)               = 14
read(14, "CET-1CEST-2,M3.5.0/02:00:00,M10."..., 68) = 45
close(14)                               = 0
getpid()                                = 11890
write(13, "<30>Sep 17 19:00:05 dnsmasq[1189"..., 72) = 72
munmap(0x4476d896, 3391088841)          = -1 EINVAL (Invalid argument)
munmap(0x69af6501, 4294967291)          = -1 EINVAL (Invalid argument)
stat64("/var/tmp/flash/dnsmasq/dnsmasq.dhcphosts", {st_mode=S_IFREG|0644, st_size=1386, ...}) = 0
open("/var/tmp/flash/dnsmasq/dnsmasq.dhcphosts", O_RDONLY|O_LARGEFILE) = 14
ioctl(14, TCGETS, 0x7fc6af98)           = -1 ENOTTY (Inappropriate ioctl for device)`
```
Dnsmasq won't even start by itself at reboot. I need to restart it in my own. All custom scripts are deactivated (onlinechanged / cron). This makes no sense.

Update:
Image without adhole --> dnsmasq starts after reboot without problems. All custom scripts still disabled.

**Reply by fda77** [reply]
Sry, io have no idea. It workd for me. Maybe dnsmasq could wrtie more (debug) logs etc

**Reply by schacka** [reply]
Well, well, I could not resist and went further into the problem. Here are my findings:

- Use "dhcp-host=" in dnsmasq-extra instead of seperate dhcp-host file
- Check dnsmasq-extra for invalid chars and use of options, check [dnsmasq-man](https://thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html)
- Happily use Adhole without (Inappropriate ioctl for device) on crashes :)

Maybe this helps others and thanks fda77 for your hints!

**Reply by fda77** [reply]
Invalid (or asian?) characters "ק|ɢ|é" in the host lists caused for me also crashes:
 https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/addhole/files/root/etc/init.d/rc.addhole#L33
